CAToolsetPerfDataCollector/PerfDataCollector.cpp

```diff
@@ -0,0 +1,306 @@
+template<size_t IndentSize>
+struct Indent
+{
+    void operator++() { buffer_.append(IndentSize, L' '); }
+    void operator--() { buffer_.resize(buffer_.size() - IndentSize); }
+    friend std::wostream& operator<<(std::wostream& out, const Indent& indent) { out << indent.buffer_; return out; }
+private:
+    std::wstring buffer_{ };
+};
+
+// Use this to automatically manage indentation level per scope for an instance of Indent<N>.
+// For example:
+//  Indent<2> indent;
+//  {
+//      Indenter _{ indent };
+//      std::wcout << indent << L"Indented by 2 spaces" << std::endl;
+//      {
+//          Indenter _{ indent };
+//          std::wcout << indent << L"Indented by 4 spaces" << std::endl;
+//      }
+//      std::wcout << indent << L"Indented by 2 spaces" << std::endl;
+//  }
+template<size_t IndentSize>
+struct Indenter
+{
+    Indenter(Indent<IndentSize>& indent) : indent_{ indent } { ++indent_; }
+    ~Indenter() { --indent_; }
+private:
+    Indent<IndentSize>& indent_;
+};
+
+// Print the summary of the performance data.
+// Prints the total execution time for each of the compiler passes and activities in CA Pass.
+// Then prints the percentage of each activity. Activities are displayed as children of
+// their parent activity. For example, AST Creation and AST Clients are displayed under CA Pass.
+// Activities' performance numbers are percentages over their immediate parent activity.
+// So, sum of percentages of all siblings should equal to 100%.
+// Also prints the number of TUs with CA Pass greater than FE Pass by over 600%, 300%, or 150%.
+void PerfDataCollector::PrintSummary()
+{
+    struct Total
+    {
+        milliseconds FEPass;                    // FrontEnd pass
+        milliseconds CAPass;                    // CodeAnalysis pass
+        milliseconds BEPass;                    // BackEnd pass
+        milliseconds ASTCreation;               // AST creation for Code Analysis
+        milliseconds ASTClients;                // COM clients for Code Analysis, including PREfast
+        milliseconds CAFunction;                // PREfast's Code Analysis of functions, including analysis of function by all of its plug-ins
+        milliseconds FPAFunction;               // PREfast's path-sensitive Code Analysis of functions (FPA = Function Path Analysis)
+        milliseconds EspXCfgBuild;              // CFG Building for functions to be used by EspXEngine and its extensions
+        milliseconds EspXAllChecks;             // Analysis of functions by EspXEngine and its extensions
+        milliseconds EspXPathSensitiveChecks;   // Path-sensitive analysis of functions by EspXEngine and its extensions
+    } total{};
+
+    size_t ignoredTUs{};
+    size_t TUsWithCAPassGE600Percent{};
+    size_t TUsWithCAPassGE300Percent{};
+    size_t TUsWithCAPassGE150Percent{};
+
+    for (const auto& [pathId, data] : perfDataPerTu)
+    {
+        // If CA Pass is less than FE Pass, it is likely CA Pass is not captured properly.
```

Comment:
How confident are we in this? I think it is not unreasonable to that sometimes the CAPass is actually faster. There might be cases where almost all functions are skipped, so the two passes do similar amount of work, and due to the variability, CA comes out a bit faster.

Also, in the future, it is possible that some parts of the FE could be skipped (like lowering to IL) when we are doing code analysis. 

I am just wondering if this is something that you hit regularly or is it just some precaution?

Reply:
This check / bailout is to exclude records where Code Analysis hits errors (for various reasons - compilation error, exceptions, etc. that only happens during CA pass) and finishes prematurely. It is really unlikely any successful CA pass to be faster than FE pass.

Reply:
In this case could this comment be expanded that we expect to see this happening due to premature compilation errors in the code analysis pass? I think this is useful information, because maybe this means that we either need a separate event for compilation errors, or if we already have one, we could rely on those to more reliably filter these cases out.

Reply:
Updated comment in the code. Regardless, it will be a good idea to log events for failed analysis, as that will always be important for us and our customers equally. The more detailed info we capture, the better, as we do not want to spend time to find what failed. We should spend our time to find out why. Let's look into how we can capture these in a future release. The earlier, the better.